File: array_c_plus_plus/55.cpp
#include<vector>
#include<iostream>

using namespace std; 

class Solution {
public:
    bool canJump(vector<int>& nums) {
        bool result=false;
        if(nums.size() == 1) return true;
        m(nums, nums.size()-2, nums.size()-1, result);
        return result;
    }

    void m(vector<int>& nums, int start, int end, bool& result){
        if(start < 0) return;
        if(nums[0] >= end - start && start == 0){
            result = true;
            return;
        }else if(nums[start] >= end - start){
            m(nums, start-1, start, result);
            return;
        }else{
            m(nums, start-1, end, result);
        }
    }
};

int main(){
    int a[] = {2, 3, 1, 1, 4};
    vector<int> nums(a, a+1);
    
    Solution s;
    bool result = s.canJump(nums);
    cout<<result<<endl;
    return -1;
}

File: DP/55.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2020-01-05 18:46:28
'''
import sys
import os

sys.path.append("..")

from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        
        is_jump = [False for _ in range(len(nums))]
        
        if len(nums) == 0:
            return False

        is_jump[0] = True

        for i in range(1, len(nums)):
            count = 1
            for j in list(range(0, i))[::-1]:
                if is_jump[j] and nums[j] >= count:
                    is_jump[i] = True
                    break
                count += 1
        
        return is_jump[-1]

if __name__ == "__main__":
    examples = [[2,3,1,1,4], [3,2,1,0,4], [2, 0, 3, 1, 1, 1, 0]]

    solu = Solution()

    for ex in examples:
        print(solu.canJump(ex))